File: Problem3.cpp
/**
 * Given the root to a binary tree, implement serialize(root), which serializes the tree into a string, 
 * and deserialize(s), which deserializes the string back into the tree.
 *  For example, given the following Node class
 * class Node:
 *   def __init__(self, val, left=None, right=None):
 *       self.val = val
 *       self.left = left
 *       self.right = right
 * The following test should pass:
 *
 * node = Node('root', Node('left', Node('left.left')), Node('right'))
 * assert deserialize(serialize(node)).left.left.val == 'left.left'
 *
 *
 *
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root)
            return "#";
        return to_string(root->val) + "," + serialize(root->left) + "," + serialize(root->right);
    }
    
    int helper(string &data)
    {
        int pos = data.find(',');
        int v = stoi(data.substr(0, pos));
        data = data.substr(pos+1);
        return v;
    }
    
    TreeNode* deserializeUtil(string &data) {
        if (data[0] == '#')
        {
            if (data.size() > 1) data = data.substr(2);
            return NULL;
        }
        
        TreeNode* root = new TreeNode(helper(data));
        root->left = deserializeUtil(data);
        root->right = deserializeUtil(data);
        
        return root;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        return deserializeUtil(data);
    }
    
    
};

File: Problem3.java
//Given the root to a binary tree, implement serialize(root),
//which serializes the tree into a string, and deserialize(s),
//which deserializes the string back into the tree

import java.util.*;

/*TreeNode has key, left right children pointers*/
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class BinaryTree {
	TreeNode root;
	
	//encode tree to single string - serializes!
	public static String serialize(TreeNode root)
	{
		if (root == null) {
			return null;
		}
		Stack<TreeNode> s = new Stack<>(); //Stack is LIFO
		s.push(root);
		
		List<String> l = new ArrayList<>();
		while (!s.isEmpty()) {
			TreeNode t = s.pop(); //return element and remove
			if (t == null) {
				//store marker
				l.add("#");
			} else {
				//store current node and recur for child
				l.add("" + t.val);
				s.push(t.right);
				s.push(t.left);
			}
		}
		return String.join(",", l);
	}
	
	static int t;
	
	//Decode encoded data back to a tree
	public static TreeNode deserialize(String data)
	{
		if (data == null)
			return null;
		t = 0;
		String[] arr = data.split(",");
		return helper(arr);
	}
	
	public static TreeNode helper(String[] arr)
	{
		if (arr[t].equals("#"))
			return null;
		//create node with item and recur for child
		TreeNode root = new TreeNode(Integer.parseInt(arr[t]));
		t++;
		root.left = helper(arr);
		t++;
		root.right = helper(arr);
		return root;
	}
	
	//in order traverse for testing tree
	static void inorder(TreeNode root)
	{
		if (root != null) {
			inorder(root.left);
			System.out.print(root.val + " ");
			inorder(root.right);
		}
	}
	
	/*Driver to test*/
	public static void main(String args[])
	{
		BinaryTree tree = new BinaryTree();
		tree.root = new TreeNode(20);
		tree.root.left = new TreeNode(8);
		tree.root.right = new TreeNode(22);
		tree.root.left.left = new TreeNode(4);
		tree.root.left.right = new TreeNode(12);
		tree.root.left.right.left = new TreeNode(10);
		tree.root.left.right.right = new TreeNode(14);
		
		String serialized = serialize(tree.root);
		System.out.println("Serialized view of the tree:");
		System.out.println(serialized);
		System.out.println();
		
		//deserialize
		TreeNode t = deserialize(serialized);
		
		System.out.println("Inorder Traversal of tree constructed from serialized String");
		inorder(t);
	}
}
